core/storage/gcp_native_chunk_manager.go

```diff
@@ -0,0 +1,472 @@
+
+	client, err = storage.NewClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create GCS client: %v", err)
+	}
+
+	if config.BucketName == "" {
+		return nil, fmt.Errorf("invalid empty bucket name")
+	}
+	// Check bucket validity
+	checkBucketFn := func() error {
+		bucket := client.Bucket(config.BucketName)
+		_, err := bucket.Attrs(ctx)
+		if err == storage.ErrBucketNotExist && config.CreateBucket {
+			log.Info("gcs bucket does not exist, create bucket.", zap.String("bucket name", config.BucketName))
+			err = client.Bucket(config.BucketName).Create(ctx, projectId, nil)
+			if err != nil {
+				return fmt.Errorf("failed to create bucket: %v", err)
+			}
+			return nil
+		}
+		return err
+	}
+	err = retry.Do(ctx, checkBucketFn, retry.Attempts(CheckBucketRetryAttempts))
+	if err != nil {
+		return nil, fmt.Errorf("bucket check failed: %v", err)
+	}
+
+	log.Info("GCP native chunk manager init success.", zap.String("bucketname", config.BucketName))
+	return &GCPNativeChunkManager{
+		client: client,
+		config: config,
+	}, nil
+}
+
+func getProjectId(credentialPath string) (string, error) {
+	if credentialPath == "" {
+		return "", errors.New("the path to the JSON file is empty")
+	}
+
+	// Read the credentials file
+	jsonData, err := os.ReadFile(credentialPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read credentials file: %v", err)
+	}
+
+	var data map[string]interface{}
+	if err := json.Unmarshal(jsonData, &data); err != nil {
+		return "", errors.New("failed to parse Google Cloud credentials as JSON")
+	}
+
+	propertyValue, ok := data["project_id"]
+	if !ok {
+		return "", errors.New("project_id doesn't exist")
+	}
+
+	projectId := fmt.Sprintf("%v", propertyValue)
+	return projectId, nil
+}
+
+// Write writes data to filePath in the specified GCS bucket
+func (gcm *GCPNativeChunkManager) Write(ctx context.Context, bucketName string, filePath string, content []byte) error {
+	wc := gcm.client.Bucket(bucketName).Object(filePath).NewWriter(ctx)
+	defer func() {
+		if err := wc.Close(); err != nil {
+			log.Error("Failed to close writer", zap.Error(err))
+		}
+	}()
+
+	_, err := wc.Write(content)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Exist checks whether chunk is saved to GCS storage.
+func (gcm *GCPNativeChunkManager) Exist(ctx context.Context, bucketName string, filePath string) (bool, error) {
+	query := &storage.Query{
+		Prefix:    filePath,
+		Delimiter: "/",
+	}
+	it := gcm.client.Bucket(bucketName).Objects(ctx, query)
+	_, err := it.Next()
+
+	if err == iterator.Done {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// Read reads the storage data if exists.
+func (gcm *GCPNativeChunkManager) Read(ctx context.Context, bucketName string, filePath string) ([]byte, error) {
+	rc, err := gcm.client.Bucket(bucketName).Object(filePath).NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
```

Comment:
Can we use defer func() here to make constant with write function?

Reply:
yes sure